### ISSUE TITLE
Github PR workflow support similar to Atlantis

### ISSUE THREAD

**Issue by sstarcher** [roboll/helmfile#386]
A workflow similar to Atlantis for terraform would be a killer feature for helmfile - https://github.com/runatlantis/atlantis/

**Comment by warrenseymour** [comment]
If you'll permit me to shamelessly plug my own wares, I created a tool for exactly this purpose, which you may find here - https://github.com/fountainheadtechnologies/branch-officer

I'm using it successfully for internal projects at the moment, and would be delighted to get feedback from 'real' users.

**Comment by manics** [comment]
@warrenseymour Is https://github.com/fountainheadtechnologies/branch-officer a private repo? I get a 404.

**Comment by warrenseymour** [comment]
@manics Hey, thanks for pointing that out, it was indeed still set to 'private', which I've now rectified. I also forgot to mention that it's been published to NPM - https://www.npmjs.com/package/@fountainhead/branch-officer

Apologies to the Go-heads for writing it in JavaScript, but that's what I'm most productive in at this time.

**Comment by mumoshu** [comment]
I was looking for something like atlantis for integration with helmfile, with no luck.

@warrenseymour Thanks for sharing your awesome work! So, perhaps we could wrap branch-officer with a kind of github webhook handler written in e.g. AWS lambda or alike to DIY build atlantis?

**Comment by warrenseymour** [comment]
@mumoshu Indeed -- personally, I'm using branch-officer with CircleCI to build PRs, but there's no reason why you couldn't achieve a similar effect by invoking it via a Lambda-based webhook.

The only thing to be mindful of is that branch-officer is currently hard-coded to expect a `GITHUB_TOKEN` environment variable. CircleCI provides this automatically at build-time, so you'd need to define this variable in your Lambda configuration yourself (and take the relevant steps to make sure that token is kept 'secret'.)

**Comment by osterman** [comment]
For the record, you can use `atlantis` with `helmfile` =)

**Comment by osterman** [comment]
@mumoshu Exactly! ...provided it has a kube context.

**Comment by mumoshu** [comment]
@osterman Awesome. Really excited to see it ☺️ 

Just curious, but how do you deal with a more complex branching strategy with helmfile + atlantis?

For example. when PRs to the `develop` branch are deployable to the `staging` env and PRs to `master` branch are deployable to the `prod` env, how do you setup atlantis+helmfile?

I firstly came up with a single `atlantis` instance and a directory-per-environment setup like this:

```
.
├── atlantis.yaml
└── deploy
    ├── environments
    │   ├── prod
    │   │   └── values.yaml
    │   └── staging
    │       └── values.yaml
    └── helmfile.yaml
```

But this doesn't prevent atlantis from running `prod` deployments from a PR to the `develop` branch.

I also get to know about https://github.com/cloudposse/atlantis/pull/13 but even with it there seems no way to prevent the `prod` atlantis intance from responding to PRs to the `develop` branch.

**Comment by sstarcher** [comment]
I don't believe atlantis has any support for branches.  Ya, so as is Atlantis could run vs helmfile you just lose things like autodiscovery and need to call out your directory structure.

**Comment by mumoshu** [comment]
I was just reading https://github.com/runatlantis/atlantis/pull/310 along with the @osterman's comment there.

I think we would need:

- A custom wake word, so that commenting on a github PR `helmfile apply` instead of `atlantis apply` triggers the atlantis instance.
- Custom workflows, so that commenting on `helmfile diff` instead of `helmfile plan` works
- A branch restriction feature, so that we can bring up separate atlantis instances per environments for gitflow-flavored projects

Please add your ideas to the list so that we can better shape what we need!

**Comment by mumoshu** [comment]
Regarding the autodiscovery feat, I assume you're talking about helmfile's convention to use the `helmfile.yaml` at the current directory(which is usually the project root), right?

I do believe it is a great nature of helmfile, and hopefully helmfile could include an out-of-box support for atlantis-like features. Not sure it is feasible to bring up our own, though.

@osterman Would there be any chance to maintain the fork of atlantis together, so that we can generalize atlantis to also support helmfile's use-cases, and add missing features?
